### ISSUE TITLE
ndt7: goroutine leak at the end of a measurement

### ISSUE THREAD

**Issue by robertodauria** [m-lab/ndt-server#361]
This was initially discovered by @phillipa by noticing that the difference between `EndTime` and `StartTime` in many ndt7 BigQuery rows is incompatible with the `MaxRuntime` (15s) parameter of ndt7 measurements. Upon further research, I have found some measurements that appear to have lasted 40+ days.

The measurer goroutine (in the `ServerMeasurements` field) ends after 10s and the `MeanClientMbps` in BigQuery is computed from those measurements -- so it appears to be correct -- but this is still an issue when doing data analysis (e.g. computing the measurement duration distribution).

For this to happen, the deferred function that sets `EndTime` at https://github.com/m-lab/ndt-server/blob/master/ndt7/upload/sender/sender.go#L47 must have been called late.

I've dumped the goroutines on a running production server (mlab1-del01) to check what's actually happening, and indeed there are receiver goroutines the have been stuck in `NextReader()` for minutes:
```

goroutine 2443141 [syscall, 9 minutes]:
syscall.Syscall(0x0, 0x23, 0xc00de67f5c, 0xc5ba, 0x4e5265, 0x7f2febd28090, 0x72)
	/usr/local/go/src/syscall/asm_linux_amd64.s:20 +0x5
syscall.read(0x23, 0xc00de67f5c, 0xc5ba, 0xc5ba, 0x0, 0x0, 0x0)
	/usr/local/go/src/syscall/zsyscall_linux_amd64.go:686 +0x5a
syscall.Read(...)
	/usr/local/go/src/syscall/syscall_unix.go:187
internal/poll.ignoringEINTRIO(...)
	/usr/local/go/src/internal/poll/fd_unix.go:581
internal/poll.(*FD).Read(0xc007c99000, 0xc00de67f5c, 0xc5ba, 0xc5ba, 0x0, 0x0, 0x0)
	/usr/local/go/src/internal/poll/fd_unix.go:162 +0x145
net.(*netFD).Read(0xc007c99000, 0xc00de67f5c, 0xc5ba, 0xc5ba, 0x5a8, 0x0, 0x0)
	/usr/local/go/src/net/fd_posix.go:55 +0x4f
net.(*conn).Read(0xc00026a4b8, 0xc00de67f5c, 0xc5ba, 0xc5ba, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:183 +0x91
crypto/tls.(*atLeastReader).Read(0xc002260a08, 0xc00de67f5c, 0xc5ba, 0xc5ba, 0x5a8, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:776 +0x63
bytes.(*Buffer).ReadFrom(0xc009a9e5f8, 0xa7d900, 0xc002260a08, 0x40c065, 0x955d00, 0x9a8340)
	/usr/local/go/src/bytes/buffer.go:204 +0xbe
crypto/tls.(*Conn).readFromUntil(0xc009a9e380, 0x7f2fec220898, 0xc00228f4c0, 0x4016, 0xc00228f4c0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:798 +0xf3
crypto/tls.(*Conn).readRecordOrCCS(0xc009a9e380, 0x0, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:650 +0x771
crypto/tls.(*Conn).readRecord(...)
	/usr/local/go/src/crypto/tls/conn.go:573
crypto/tls.(*Conn).Read(0xc009a9e380, 0xc0149b6000, 0x100000, 0x100000, 0x0, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:1276 +0x165
bufio.(*Reader).Read(0xc002d79200, 0xc01221a000, 0x2000, 0x2000, 0x1ff2, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:227 +0x222
io.(*LimitedReader).Read(0xc0022609f0, 0xc01221a000, 0x2000, 0x2000, 0x1ff2, 0x0, 0x0)
	/usr/local/go/src/io/io.go:473 +0x63
io.discard.ReadFrom(0xa7dfc0, 0xc0022609f0, 0x9727c0, 0x972701, 0x7f2febd099f8)
	/usr/local/go/src/io/io.go:598 +0x95
io.copyBuffer(0xa7eb40, 0xdae420, 0xa7dfc0, 0xc0022609f0, 0x0, 0x0, 0x0, 0xc0149b600a, 0x0, 0x0)
	/usr/local/go/src/io/io.go:409 +0x357
io.Copy(...)
	/usr/local/go/src/io/io.go:382
io.CopyN(0xa7eb40, 0xdae420, 0xa7d760, 0xc002d79200, 0x10000, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:358 +0x9a
github.com/gorilla/websocket.(*Conn).advanceFrame(0xc0098622c0, 0x0, 0x0, 0x0)
	/go/pkg/mod/github.com/gorilla/websocket@v1.4.2/conn.go:791 +0xe34
github.com/gorilla/websocket.(*Conn).NextReader(0xc0098622c0, 0x0, 0x0, 0xc000d801f0, 0x0, 0x0)
	/go/pkg/mod/github.com/gorilla/websocket@v1.4.2/conn.go:980 +0x8f
github.com/m-lab/ndt-server/ndt7/receiver.start(0xa87110, 0xc0100fb8c0, 0xc0098622c0, 0x1, 0xc007dc2aa0)
	/go/src/github.com/m-lab/ndt-server/ndt7/receiver/receiver.go:58 +0x4b9
github.com/m-lab/ndt-server/ndt7/receiver.StartUploadReceiverAsync.func1(0xa87110, 0xc0100fb8c0, 0xc0098622c0, 0xc007dc2aa0, 0xc01c3a8c40)
	/go/src/github.com/m-lab/ndt-server/ndt7/receiver/receiver.go:121 +0x52
created by github.com/m-lab/ndt-server/ndt7/receiver.StartUploadReceiverAsync
	/go/src/github.com/m-lab/ndt-server/ndt7/receiver/receiver.go:120 +0x85
```

We do set a Read timeout on the connection at https://github.com/m-lab/ndt-server/blob/master/ndt7/receiver/receiver.go#L37 but either the read deadline does not work as we thought, or this is expected behavior and we should force-close the connection after the timeout on a separate goroutine rather than relying on Read() deadlines.

There seems to be exactly one place where we force-close the connection in the ndt7 package, and it's https://github.com/m-lab/ndt-server/blob/6aee5ccbcbf4e47eeead8ad52e91aaeb174d72fc/ndt7/handler/handler.go#L71. However, since it's in a deferred function, it does not run until the receiver has returned, so it does not help when the receiver goroutine is stuck.

Something similar is also happening with ndt7 downloads. e.g.:
```
goroutine 2447240 [syscall, 4 minutes]:
syscall.Syscall(0x1, 0x87, 0xc004a83800, 0x4016, 0x4e5265, 0x7f2febd27ec0, 0x77)
	/usr/local/go/src/syscall/asm_linux_amd64.s:20 +0x5
syscall.write(0x87, 0xc004a83800, 0x4016, 0x4800, 0x0, 0x4001, 0x47fb)
	/usr/local/go/src/syscall/zsyscall_linux_amd64.go:914 +0x5a
syscall.Write(...)
	/usr/local/go/src/syscall/syscall_unix.go:212
internal/poll.ignoringEINTRIO(...)
	/usr/local/go/src/internal/poll/fd_unix.go:581
internal/poll.(*FD).Write(0xc000192780, 0xc004a83800, 0x4016, 0x4800, 0x0, 0x0, 0x0)
	/usr/local/go/src/internal/poll/fd_unix.go:274 +0x13a
net.(*netFD).Write(0xc000192780, 0xc004a83800, 0x4016, 0x4800, 0x1a, 0x0, 0x8)
	/usr/local/go/src/net/fd_posix.go:73 +0x4f
net.(*conn).Write(0xc0000101f8, 0xc004a83800, 0x4016, 0x4800, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:195 +0x8e
crypto/tls.(*Conn).write(0xc009a9f500, 0xc004a83800, 0x4016, 0x4800, 0xc000ef4000, 0x4000, 0x5008)
	/usr/local/go/src/crypto/tls/conn.go:911 +0x162
crypto/tls.(*Conn).writeRecordLocked(0xc009a9f500, 0x17, 0xc000ef4000, 0x4004, 0x5008, 0x0, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:979 +0x352
crypto/tls.(*Conn).Write(0xc009a9f500, 0xc000ef4000, 0x4004, 0x5008, 0x0, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:1150 +0x19a
github.com/gorilla/websocket.(*Conn).write(0xc01941c580, 0x2, 0xc0afbb4d0e3997e9, 0x145b945f82a67, 0xd7f2c0, 0xc000ef4000, 0x4004, 0x5008, 0x0, 0x0, ...)
	/go/pkg/mod/github.com/gorilla/websocket@v1.4.2/conn.go:391 +0x153
github.com/gorilla/websocket.(*Conn).WritePreparedMessage(0xc01941c580, 0xc001cb31d0, 0xc01b432300, 0x0)
	/go/pkg/mod/github.com/gorilla/websocket@v1.4.2/conn.go:741 +0x116
github.com/m-lab/ndt-server/ndt7/download/sender.Start(0xa871b8, 0xc010592780, 0xc01941c580, 0xc0042180a0, 0x0, 0x0)
	/go/src/github.com/m-lab/ndt-server/ndt7/download/sender/sender.go:94 +0xc48
github.com/m-lab/ndt-server/ndt7/download.Do(0xa871b8, 0xc010592780, 0xc01941c580, 0xc0042180a0, 0xa89ca0, 0xc0003e2060)
	/go/src/github.com/m-lab/ndt-server/ndt7/download/download.go:27 +0x8d
github.com/m-lab/ndt-server/ndt7/handler.Handler.runMeasurement(0x7fffa79aca46, 0xe, 0x9df815, 0x4, 0x9df610, 0x3, 0xc000234440, 0x2, 0x2, 0x9e0e7a, ...)
	/go/src/github.com/m-lab/ndt-server/ndt7/handler/handler.go:99 +0x4bc
github.com/m-lab/ndt-server/ndt7/handler.Handler.Download(...)
	/go/src/github.com/m-lab/ndt-server/ndt7/handler/handler.go:53
net/http.HandlerFunc.ServeHTTP(0xc0002205f0, 0xa863b0, 0xc00db90500, 0xc009f1a800)
	/usr/local/go/src/net/http/server.go:2049 +0x44
net/http.(*ServeMux).ServeHTTP(0xc0002345c0, 0xa863b0, 0xc00db90500, 0xc009f1a800)
	/usr/local/go/src/net/http/server.go:2428 +0x1ad
github.com/gorilla/handlers.loggingHandler.ServeHTTP(0xa7e480, 0xc00013e010, 0xa7e200, 0xc0002345c0, 0xa05318, 0xa85ae0, 0xc0000200e0, 0xc009f1a800)
	/go/pkg/mod/github.com/gorilla/handlers@v1.5.1/logging.go:47 +0xe6
github.com/m-lab/access/controller.(*TxController).Limit.func1(0xa85ae0, 0xc0000200e0, 0xc009f1a800)
	/go/pkg/mod/github.com/m-lab/access@v0.0.9/controller/tx.go:130 +0x129
net/http.HandlerFunc.ServeHTTP(0xc000222400, 0xa85ae0, 0xc0000200e0, 0xc009f1a800)
	/usr/local/go/src/net/http/server.go:2049 +0x44
github.com/m-lab/access/controller.(*TokenController).Limit.func1(0xa85ae0, 0xc0000200e0, 0xc009f1a700)
	/go/pkg/mod/github.com/m-lab/access@v0.0.9/controller/token.go:100 +0x9b
net/http.HandlerFunc.ServeHTTP(0xc000222420, 0xa85ae0, 0xc0000200e0, 0xc009f1a700)
	/usr/local/go/src/net/http/server.go:2049 +0x44
net/http.serverHandler.ServeHTTP(0xc000254380, 0xa85ae0, 0xc0000200e0, 0xc009f1a700)
	/usr/local/go/src/net/http/server.go:2867 +0xa3
net/http.(*conn).serve(0xc007dc23c0, 0xa871b8, 0xc00e71c140)
	/usr/local/go/src/net/http/server.go:1932 +0x8cd
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:2993 +0x39b
```

**Comment by robertodauria** [comment]
It's still unclear what exactly causes the original issue. AFAICT, Go's `net` package creates sockets in non-blocking mode and implements its own polling loop which waits for data to be available and makes `Read()` blocking. However, by looking at production servers' goroutine dumps, the receiver is definitely blocked on a (supposedly non-blocking) read syscall implemented in asm, not on Go's own polling loop. It's possible that there are TCP states where this makes sense, but I don't know enough about that to draw a conclusion. Spot-checking a few pcaps, they either show no one terminating the connection (no FINs sent) or an out-of-the-blue FIN/ACK from the client, which would make sense if we sent a FIN first, but there's no trace of that in the pcap. It might be a misbehaving middlebox that makes the client believe we closed the connection.
